### PR TITLE
feat(dsd-fp2): ASCOM CoverCalibrator driver for Deep Sky Dad FP2

### PR DESCRIPTION
**Pull request by ivonnyssen**
## Summary

- New `dsd-fp2` service — ASCOM Alpaca **CoverCalibrator** driver for the Deep Sky Dad Flat Panel 2, port **11119**.
- First adopter of `rusty-photon-shared-transport` (PR #269): the FP2's bracketed-ASCII protocol is plugged in as `Fp2Codec`; refcounting, request arbitration, the 500 ms poll loop, and rollback-on-handshake-failure all come from the shared crate.
- Verified end-to-end against the live FP2 panel; pairs with `calibrator-flats` without any orchestrator changes.

## What's in the box

- `services/dsd-fp2/src/` — 10 source files: `codec.rs`, `transport.rs`, `manager.rs`, `device.rs`, `mock.rs`, `protocol.rs`, `config.rs`, `error.rs`, `lib.rs`, `main.rs`.
- `services/dsd-fp2/tests/` — 11 BDD scenarios across 3 feature files (`connection_lifecycle`, `cover_control`, `calibrator_control`) plus a deliberately-unwired ConformU integration scaffold.
- `docs/services/dsd-fp2.md` — design doc: architecture, protocol reference, ASCOM mapping, lifecycle, MVP scope, known limitations.
- Workspace `Cargo.toml` registers the new member; root `README.md` adds the row and coverage badge.

## Hardware verification

Driven with `curl` against the running service while watching the physical panel:

| ASCOM action | Wire | Physical |
|---|---|---|
| `connected = true` | `[GFRM]` ⇒ `(Board=DeepSkyDad.FP2, Version=1.0.14.2)` + cached-state seed | panel enters remote-control mode |
| `calibrator_on(500)` then `(4096)` | `[SLBR0500]`/`[SLBR4096]` + `[SLON1]` | panel illuminates dimly, then at full brightness |
| `calibrator_off` | `[SLON0]` | panel dark |
| `open_cover` | `[STRG0]` + `[SMOV]` | cover swings open (~6 s); `cover_state` Moving → Open |
| `close_cover` | `[STRG270]` + `[SMOV]` | cover swings closed (~6 s); `cover_state` Moving → Closed |
| `halt_cover` | _(no wire)_ | returns `NOT_IMPLEMENTED` per ICoverCalibratorV2 spec ("if cover movement cannot be interrupted") |
| `calibrator_on(5000)` | _(no wire)_ | returns `INVALID_VALUE` |
| `connected = false` | clean teardown | poll loop cancelled, session closed |

## ConformU

| | Issues | Errors | Configuration alerts |
|---|---|---|---|
| Mock factory | 1 | 0 | 0 |
| Real FP2 hardware | 1 | 0 | 0 |

The single issue in both runs is the same: `HaltCover` returns `MethodNotImplementedException` per the [ASCOM ICoverCalibratorV2 spec](https://ascom-standards.org/newdocs/covercalibrator.html) — ConformU 4.3's `CoverCalibratorTester` does not honour the spec's "if cover movement cannot be interrupted" qualifier. Filed upstream as [ASCOMInitiative/ConformU#30](https://github.com/ASCOMInitiative/ConformU/issues/30). The crate intentionally omits `[package.metadata.conformu]` so the rolling ConformU matrix does not pick up dsd-fp2; the integration scaffold stays for re-enabling once upstream lands.

All other 30-odd ConformU checks pass on real hardware, and all members return within their target response times (`OpenCover`/`CloseCover` initiate in ~0.1 s; the cover then transits asynchronously for ~6.1 s).

## Deferred (follow-up PRs)

- Heater control as a second ASCOM `Switch` device (FP2 exposes `[GHTT]` / `[GHTM]` / `[SHTM]`; reads are already in the poll cycle).
- Re-enable the ConformU integration test once upstream #30 lands.
- Brightness ramp / LUT for non-linear EL perceived-brightness.
- Per-filter brightness profiles surfaced to `calibrator-flats` (orchestrator change, not driver).
- Sentinel dashboard tile for FP2 heater temperature.

## Test plan

- [ ] `cargo rail run --profile commit -q` is green from a fresh checkout
- [ ] `cargo test -p dsd-fp2 --features mock --test bdd` runs 11 scenarios / 52 steps and passes
- [ ] `cargo test -p dsd-fp2 --all-features --lib` runs 78 unit tests and passes
- [ ] _(optional, requires hardware)_ `cargo run -p dsd-fp2 -- --port /dev/serial/by-id/usb-Deep_Sky_Dad_Deep_Sky_Dad_FP2_*-if00` then exercise the Alpaca endpoints via `curl http://localhost:11119/api/v1/covercalibrator/0/...`
- [ ] _(optional, requires hardware + ConformU)_ `conformu conformance http://localhost:11119/api/v1/covercalibrator/0` reports exactly 1 issue (HaltCover; see ConformU#30) and zero errors

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: services/dsd-fp2/tests/bdd/world.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        // Read the simulator state directly and update the cached snapshot.
+        let mut s = snap.write().await;
+        // Use the simulator's observable accessors plus a direct lock read
+        // for the cover angle — the device itself doesn't expose this.
+        s.light_on = Some(state.light_on().await);
+        s.brightness = Some(state.brightness().await);
+        // For motor/cover, fall through to the simulator's command path
+        // (it's already exercised by unit tests; here we just hard-code
+        // the post-move steady state since our mock completes moves
+        // instantly).
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/dsd-fp2/tests/bdd/world.rs]: Agree, and this is the root cause of all three failing CI checks (`ubuntu / stable`, `windows / bdd / dsd-fp2`, `coverage`) — they all fail with the same `cover_state` Closed vs Open mismatch on this exact assertion. My local Linux run passed by lucky race: tokio's `interval(d)` fires immediately on the first tick, so the while-open poll task happened to refresh `cover_raw` between my `refresh_cache` call and the assertion. CI runners hit the other timing. Fix: have `refresh_cache` query the mock's cover angle and derive `cover_raw` + `motor_running` directly, no longer relying on the poll task firing in time.

File: services/dsd-fp2/tests/conformu_integration.rs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    std::fs::write(
+        &conformu_settings_path,
+        serde_json::to_string_pretty(&conformu_settings)?,
+    )?;
+
+    let config = serde_json::json!({
+        "serial": {
+            "port": "/dev/mock",
+            "baud_rate": 115200,
+            "polling_interval": "60s",
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/dsd-fp2/tests/conformu_integration.rs]: Agree. The 60 s polling was copied verbatim from qhy-focuser without thinking through whether it fit a CoverCalibrator's request/response shape. Lowering to 500 ms (the production default) — the shared-transport `Connection.transport` mutex serialises the poll task against incoming requests, so there's no race risk.

File: docs/services/dsd-fp2.md [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+- **Command framing**: `[CMD...]` — opening `[`, ASCII payload, closing `]`.
+  No terminator is added.
+- **Response framing**: `(VALUE)` — opening `(`, ASCII payload, closing `)`.
+  Read until `)`. Timeout: 3 s per request.
+- **Flush** the serial buffers before each command (the INDI reference
+  driver does this with `tcflush(TCIOFLUSH)`; we replicate by clearing
+  any pending bytes before write).
+- Commands are serialised: only one outstanding `(send, recv)` pair at a
+  time, gated by a `command_lock` mutex in `SerialManager`. This matches
+  the `qhy-focuser` and `ppba-driver` pattern.
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply docs/services/dsd-fp2.md]: Agree, leftover from the pre-refactor draft. Will rewrite the four locations to describe the actual shape (`Codec` plus per-`Connection` request arbitration via `Connection.transport`'s Mutex in `rusty-photon-shared-transport`) and drop the pre-write drain claim — `SerialFrameTransport` doesn't do that and we don't need it for the FP2 wire format.

File: services/dsd-fp2/tests/bdd.rs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+//! BDD entry point for the dsd-fp2 driver.
+//!
+//! Scenarios run in-process against the mock `TransportFactory`; no
+//! subprocess spawn, so `bdd_infra::bdd_main!` is unnecessary (same
+//! pattern as the qhy-focuser BDD suite for in-process tests).
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/dsd-fp2/tests/bdd.rs]: Agree, the comparison is wrong. qhy-focuser's BDD spawns the binary via `bdd_infra::bdd_main!`; ours is genuinely in-process against the mock factory. Will rewrite to describe the in-process approach without the false comparison.

File: services/dsd-fp2/src/device.rs [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+        Err(ASCOMError::new(
+            ASCOMErrorCode::NOT_IMPLEMENTED,
+            "HaltCover not implemented: FP2 firmware cannot interrupt an in-progress cover \
+             movement. Per ICoverCalibratorV2, HaltCover MUST throw MethodNotImplementedException \
+             when cover movement cannot be interrupted.",
+        ))
+    }
+
+    async fn calibrator_on(&self, brightness: u32) -> ASCOMResult<()> {
+        let value = FlatPanelManager::validate_brightness(brightness).map_err(Self::ascom_err)?;
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/dsd-fp2/src/device.rs]: Agree. Will validate `calibrator_on(brightness)` against the effective max — `min(config.max_brightness, MAX_BRIGHTNESS as u32)` — so a config-capped value rejects writes that are within the hardware limit but above the configured cap. Plus a unit test pinning the contract.

**Comment by ivonnyssen** [comment]
Reviewing the 3 low-confidence comments in the Copilot review body — all accurate:

* line 380 (`MockSerialPortFactory`) — stale; the actual type is `MockTransportFactory`.
* line 388 (`serial_manager.rs` reference) — stale; state-derivation tests live in `device.rs` (`derive_cover_state` / `derive_calibrator_state`) and refcount edge cases live in `manager.rs`'s mock-tests + the shared-transport crate's own integration suite.
* line 315 (`InvalidBrightness(u32)`) — stale; the actual variant is `InvalidValue(String)`.

Will fix all three in the same commit as the inline-comment fixes.

File: services/dsd-fp2/src/device.rs [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+                    .map_err(|e| Self::ascom_err(flatten_session_error(e)))?;
+                *slot = Some(session);
+                debug!("FP2 device connected");
+            }
+            false => {
+                if let Some(session) = slot.take() {
+                    session
+                        .close()
+                        .await
+                        .map_err(|e| Self::ascom_err(DsdFp2Error::Communication(e.to_string())))?;
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/dsd-fp2/src/device.rs]: Comment correct that close I/O serialises against `connected()`/state reads. But the identical shape (`slot.take()` + `session.close().await` under the write lock) is in `qhy-focuser`, `ppba-driver`, `pa-falcon-rotator`, and `star-adventurer-gti`'s `set_connected`. I'd rather keep dsd-fp2 consistent with the established pattern and lift the lock release across all five services in a separate workspace-wide PR than diverge here. Skipping on this PR.

File: services/dsd-fp2/src/device.rs [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+        assert_eq!(err.code, ascom_alpaca::ASCOMErrorCode::INVALID_VALUE);
+        device.set_connected(false).await.unwrap();
+    }
+
+    #[tokio::test]
+    async fn open_cover_then_close_cover_change_observable_state() {
+        let (device, _) = make_device();
+        device.set_connected(true).await.unwrap();
+
+        // Default is open; close it.
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/dsd-fp2/src/device.rs]: Right — mock defaults to `cover_angle = 270` (`mock.rs:42`), which `[GOPS]` reports as `0` (the device's convention: 0=closed, 1=open) and `derive_cover_state(Some(false), Some(0))` → `Closed`. The test comment is stale. Will fix.

File: docs/services/dsd-fp2.md [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+
+`CalibratorState` is derived from cached `[GLON]`:
+
+| `[GLON]` | `CalibratorState` |
+|----------|-------------------|
+| `0`      | `Off`             |
+| `1`      | `Ready`           |
+| other    | `Unknown`         |
+| —        | `Unknown` (disconnected) |
+| parse err| `Error`           |
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply docs/services/dsd-fp2.md]: Agree — `CoverStatus::Error` / `CalibratorStatus::Error` are referenced only in the BDD step parser (`world.rs:132`,`:144`); neither `derive_cover_state` nor `derive_calibrator_state` ever returns them. Codec-layer parse errors propagate as `DsdFp2Error::MalformedResponse` through the wire path, not as a state-derivation error. Will drop the spurious `parse err → Error` rows from both tables.

File: docs/services/dsd-fp2.md [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+are converted to `ASCOMError` for protocol boundaries:
+
+| Internal error               | ASCOM error code                       |
+|------------------------------|-----------------------------------------|
+| `NotConnected`               | `NOT_CONNECTED`                        |
+| `InvalidValue(String)`       | `INVALID_VALUE`                        |
+| `Timeout(_)`                 | `UNSPECIFIED` with `Operation timed out` message |
+| `Communication(_)`           | `UNSPECIFIED`                          |
+| `MalformedResponse(_)`       | `UNSPECIFIED`                          |
+| `Io(_)`                      | `UNSPECIFIED`                          |
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply docs/services/dsd-fp2.md]: Agree — `DsdFp2Error::to_ascom_error()` routes `Timeout` / `Communication` / `MalformedResponse` / `Io` / `SerialPort` through `ASCOMError::invalid_operation(...)`, which is `INVALID_OPERATION`, not `UNSPECIFIED`. Will correct the four rows.

File: docs/services/dsd-fp2.md [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+
+Commands not in the above table are unused by this driver. The FP1
+variant of the protocol uses some additional commands (e.g. `[GPOS]`-only
+state) that are exercised here only as a fallback / liveness ping.
+
+### Ping / Liveness
+
+The handshake on connect issues `[GFRM]`. A non-empty response identifying
+"DeepSkyDad.FP2" succeeds; anything else fails the connection with a
+clear error. The firmware string is cached for `driver_info()`.
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply docs/services/dsd-fp2.md]: Agree — `driver_info()` returns a static string (`device.rs:113`); the handshake-cached `firmware_board`/`firmware_version` are populated for future use but never read out through ASCOM. I'd rather update the doc than wire firmware into `driver_info()` — by convention that field is the *driver* identity, not the connected hardware's (firmware identity belongs in `device_info` or an explicit accessor). Will fix the docs.

File: services/dsd-fp2/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+            .await?
+    };
+
+    #[cfg(not(feature = "mock"))]
+    let bound = ServerBuilder::new().with_config(config).build().await?;
+
+    tokio::select! {
+        result = bound.start() => { result?; }
+        () = shutdown_signal() => { info!("Shutting down"); }
+    }
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/dsd-fp2/src/main.rs:120]: Agree the outer `tokio::select!` cancels `bound.start()` mid-flight when the signal branch wins, bypassing the inner `with_graceful_shutdown` future inside `serve_plain`. The same shape (verbatim) is in `qhy-focuser`, `ppba-driver`, `pa-falcon-rotator`, and `star-adventurer-gti`'s `main.rs`. Want to fix uniformly across services rather than diverge dsd-fp2; skipping on this PR and filing a workspace-wide follow-up.

File: services/dsd-fp2/src/protocol.rs [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+        assert_eq!(r.body, "26.104242");
+    }
+
+    #[test]
+    fn from_frame_rejects_missing_brackets() {
+        assert!(matches!(
+            RawResponse::from_frame("OK"),
+            Err(DsdFp2Error::MalformedResponse(_))
+        ));
+        assert!(matches!(
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/dsd-fp2/src/protocol.rs:244]: Agree — the test exercises missing `(` / `)` parens (FP2's response framing), not `[` / `]` brackets (the command framing). Will rename to `from_frame_rejects_missing_parens`.

File: docs/services/dsd-fp2.md [hunk 13]
```diff
@@ -0,0 +1,462 @@
...
+| `brightness()`                | _(cached)_                              | Cached `[GLBR]`                                                                                      |
+| `max_brightness()`            | _(constant)_                            | `4096`                                                                                               |
+| `open_cover()`                | `[STRG0]` → `(OK)` ; `[SMOV]` → `(OK)`  | Asynchronous; `cover_state` reports `Moving` until polled `[GMOV]→(0)`.                              |
+| `close_cover()`               | `[STRG270]` → `(OK)` ; `[SMOV]` → `(OK)` | Same as above.                                                                                      |
+| `halt_cover()`                | — (returns `MethodNotImplementedException`) | The FP2 firmware has no halt-motion opcode; per the ASCOM spec, `HaltCover` MUST throw `MethodNotImplementedException` when cover movement cannot be interrupted. |
+| `calibrator_on(brightness)`   | `[SLBR<NNNN>]` → `(OK)` ; `[SLON1]` → `(OK)` | Clamp `brightness` to `0..=4096`; pad to 4 digits. Sending brightness even when already on keeps the call idempotent. |
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply docs/services/dsd-fp2.md:228]: Agree — table is stale. Since the validation overhaul in 5f98d8d, `max_brightness()` returns `effective_max_brightness() = min(config.max_brightness, MAX_BRIGHTNESS as u32)`, and `calibrator_on` validates against that effective cap (`device.rs:185-196`). Will update the row + the surrounding validation bullets to reflect the configurable cap.
